File: ios/Sources/MqttQuicPlugin/MQTT/MQTTTypes.swift
//
// MQTTTypes.swift
// MqttQuicPlugin
//
// MQTT packet types and constants. Matches MQTTD protocol.py / protocol_v5.py.
//

import Foundation

/// MQTT 3.1.1 / 5.0 message types (fixed header first nibble)
public enum MQTTMessageType: UInt8 {
    case CONNECT = 0x10
    case CONNACK = 0x20
    case PUBLISH = 0x30
    case PUBACK = 0x40
    case PUBREC = 0x50
    case PUBREL = 0x62
    case PUBCOMP = 0x70
    case SUBSCRIBE = 0x82
    case SUBACK = 0x90
    case UNSUBSCRIBE = 0xA2
    case UNSUBACK = 0xB0
    case PINGREQ = 0xC0
    case PINGRESP = 0xD0
    case DISCONNECT = 0xE0
}

/// CONNECT flags
public struct MQTTConnectFlags {
    public static let username: UInt8 = 0x80
    public static let password: UInt8 = 0x40
    public static let willRetain: UInt8 = 0x20
    public static let willQoS1: UInt8 = 0x08
    public static let willQoS2: UInt8 = 0x18
    public static let willFlag: UInt8 = 0x04
    public static let cleanSession: UInt8 = 0x02
    public static let reserved: UInt8 = 0x01
}

/// CONNACK return codes (MQTT 3.1.1)
public enum MQTTConnAckCode: UInt8 {
    case accepted = 0x00
    case unacceptableProtocol = 0x01
    case identifierRejected = 0x02
    case serverUnavailable = 0x03
    case badUsernamePassword = 0x04
    case notAuthorized = 0x05
}

/// Protocol levels
public struct MQTTProtocolLevel {
    public static let v311: UInt8 = 0x04
    public static let v5: UInt8 = 0x05
}

File: android/src/main/kotlin/ai/annadata/mqttquic/mqtt/MQTTTypes.kt
package ai.annadata.mqttquic.mqtt

/**
 * MQTT packet types and constants. Matches MQTTD protocol.py / protocol_v5.py.
 */

object MQTTMessageType {
    const val CONNECT: Byte = 0x10.toByte()
    const val CONNACK: Byte = 0x20.toByte()
    const val PUBLISH: Byte = 0x30.toByte()
    const val PUBACK: Byte = 0x40.toByte()
    const val PUBREC: Byte = 0x50.toByte()
    const val PUBREL: Byte = 0x62.toByte()
    const val PUBCOMP: Byte = 0x70.toByte()
    const val SUBSCRIBE: Byte = 0x82.toByte()
    const val SUBACK: Byte = 0x90.toByte()
    const val UNSUBSCRIBE: Byte = 0xA2.toByte()
    const val UNSUBACK: Byte = 0xB0.toByte()
    const val PINGREQ: Byte = 0xC0.toByte()
    const val PINGRESP: Byte = 0xD0.toByte()
    const val DISCONNECT: Byte = 0xE0.toByte()
    const val AUTH: Byte = 0xF0.toByte()
}

object MQTTConnectFlags {
    const val USERNAME: Int = 0x80
    const val PASSWORD: Int = 0x40
    const val WILL_RETAIN: Int = 0x20
    const val WILL_QOS1: Int = 0x08
    const val WILL_QOS2: Int = 0x18
    const val WILL_FLAG: Int = 0x04
    const val CLEAN_SESSION: Int = 0x02
    const val RESERVED: Int = 0x01
}

object MQTTConnAckCode {
    const val ACCEPTED: Int = 0x00
    const val UNACCEPTABLE_PROTOCOL: Int = 0x01
    const val IDENTIFIER_REJECTED: Int = 0x02
    const val SERVER_UNAVAILABLE: Int = 0x03
    const val BAD_USERNAME_PASSWORD: Int = 0x04
    const val NOT_AUTHORIZED: Int = 0x05
}

object MQTTProtocolLevel {
    const val V311: Byte = 0x04
    const val V5: Byte = 0x05
}
